cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -19,152 +22,211 @@ class UpgradeCommand < Kontena::Command
-      normalized_data = {
-        parent_name => stack_data.merge(
-          :loader => loader_class.for(stack_data['stack'])
-        )
-      }
+    # @return [Kontena::Cli::Stacks::ChangeResolver]
+    def execute
+      set_env_variables(stack_name, current_grid)
 
-      depends.each do |stack|
-        key = "#{parent_name}-#{stack['name']}"
-        normalized_data.merge!(normalize_local_data(stack.merge('parent_name' => parent_name), key))
+      old_data = spinner "Reading stack #{pastel.cyan(stack_name)} from master" do
+        gather_master_data(stack_name)
       end
-      normalized_data
-    end
 
-    def normalize_master_data(stack_name, raise_not_found = false)
-      begin
-        data = fetch_master_data(stack_name)
-      rescue Kontena::Errors::StandardError => ex
-        return nil if ex.status == 404 && !raise_not_found
-        raise ex
+      new_data = spinner "Parsing #{pastel.cyan(source)}" do
+        loader.flat_dependencies(
+          stack_name,
+          variables: values_from_options
+        )
       end
-      depends = data.delete('children') || []
 
-      normalized_data = { stack_name => data }
+      changes = process_data(old_data, new_data)
 
-      return normalized_data if skip_dependencies?
+      display_report(changes)
 
-      depends.each do |stack|
-        normalized_data.merge!(normalize_master_data(stack['name']))
-      end
-      normalized_data
+      return if dry_run?
+
+      get_confirmation(changes)
+
+      run_removes(changes.removed_stacks)
+
+      deployable_stacks = []
+      deployable_stacks.concat run_installs(changes)
+      deployable_stacks.concat run_upgrades(changes)
+
+      run_deploys(deployable_stacks) if deploy?
+
+      changes
     end
 
-    def merge_data(local_data, remote_data)
-      merged = {}
-      unless local_data.nil? || local_data.empty?
-        local_data.each do |key, data|
-          merged[key] ||= {}
-          merged[key][:local] = data
-        end
+    private
+
+    # Recursively fetch master data in StackFileLoader#flat_dependencies format
+    # @return [Hash{string => Hash}] stackname => hash
+    def gather_master_data(stackname)
+      response = fetch_master_data(stackname)
+      children = response.delete('children') || []
+      result = { stackname => { stack_data: response } }
+      children.each do |child|
+        result.merge!(gather_master_data(child['name']))
       end
-      unless remote_data.nil? || remote_data.empty?
-        remote_data.each do |key, data|
-          merged[key] ||= {}
-          merged[key][:remote] = data
-        end
+      result
+    end
+
+    # Preprocess data and return a ChangeResolver
+    # @param old_data [Hash] data from master
+    # @param new_data [Hash] data from files
+    # @return [Kontena::Cli::Stacks::ChangeRsolver]
+    def process_data(old_data, new_data)
+      caret "Analyzing upgrade" # not a spinner, because the stack executes can create prompts
```

Comment:
Surprising ordering here, because the `Analyzing upgrade` comes before the stack variable prompts. It can't analyze the upgrade if it doesn't have the new stack variable values yet?

```
> Analyzing upgrade ..
> Redis version : (3.0) 
```